src_lockfree/spsc_queue.ml

```diff
@@ -24,82 +24,79 @@
 type 'a t = {
   array : 'a Option.t Array.t;
   tail : int Atomic.t;
+  tail_cache : int ref;
   head : int Atomic.t;
-  mask : int;
+  head_cache : int ref;
 }
 
 exception Full
 
 let create ~size_exponent =
-  let size = Int.shift_left 1 size_exponent in
-  {
-    head = Atomic.make 0;
-    tail = Atomic.make 0;
-    mask = size - 1;
-    array = Array.init size (fun _ -> None);
-  }
+  if size_exponent < 0 || Sys.int_size - 2 < size_exponent then
+    invalid_arg "size_exponent out of range";
+  let size = 1 lsl size_exponent in
+  let array = Array.make size None in
+  let tail = Atomic.make_contended 0 in
+  let tail_cache = ref 0 in
+  let head = Atomic.make_contended 0 in
+  let head_cache = ref 0 in
+  { array; tail; tail_cache; head; head_cache }
 
-let push { array; head; tail; mask; _ } element =
-  let size = mask + 1 in
-  let head_val = Atomic.get head in
-  let tail_val = Atomic.get tail in
-  if head_val + size == tail_val then raise Full
-  else (
-    Array.set array (tail_val land mask) (Some element);
-    Atomic.set tail (tail_val + 1))
+type _ mono = Unit : unit mono | Bool : bool mono
 
-let try_push { array; head; tail; mask; _ } element =
-  let size = mask + 1 in
-  let head_val = Atomic.get head in
-  let tail_val = Atomic.get tail in
-  if head_val + size == tail_val then false
-  else (
-    Array.set array (tail_val land mask) (Some element);
-    Atomic.set tail (tail_val + 1);
-    true)
+let push_as (type r) t element (mono : r mono) : r =
+  let size = Array.length t.array in
+  let tail = Atomic.get t.tail in
+  let head_cache = !(t.head_cache) in
+  if
+    head_cache == tail - size
+    &&
+    let head = Atomic.get t.head in
+    t.head_cache := head;
+    head == head_cache
+  then match mono with Unit -> raise Full | Bool -> false
+  else begin
+    Array.unsafe_set t.array (tail land (size - 1)) (Some element);
+    Atomic.incr t.tail;
+    match mono with Unit -> () | Bool -> true
+  end
+
+let push t element = push_as t element Unit
```

Comment:
This wasn't changed/introduced in this PR, but I wonder if we should start using the convention that an operation like this raises an exception to indicate an expected condition (rather than an error) would be named with an `_exn` suffix.  IOW, `pop -> pop_exn` and `push -> push_exn`.